dev/Scroller/SnapPointWrapper.cpp

```diff
@@ -0,0 +1,161 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+#include "common.h"
+#include "SnapPointWrapper.h"
+
+template<typename T>
+SnapPointWrapper<T>::SnapPointWrapper(T const& snapPoint)
+    : m_snapPoint(snapPoint)
+{
+}
+
+#ifdef _DEBUG
+template<typename T>
+SnapPointWrapper<T>::~SnapPointWrapper()
+{
+}
+#endif //_DEBUG
+
+template<typename T>
+T SnapPointWrapper<T>::SnapPoint() const
+{
+    return m_snapPoint;
+}
+
+template<typename T>
+std::tuple<double, double> SnapPointWrapper<T>::ActualApplicableZone() const
+{
+    return m_actualApplicableZone;
+}
+
+template<typename T>
+int SnapPointWrapper<T>::CombinationCount() const
+{
+    return m_combinationCount;
+}
+
+template<typename T>
+winrt::ExpressionAnimation SnapPointWrapper<T>::CreateRestingPointExpression(
+    winrt::Compositor const& compositor,
+    winrt::hstring const& target,
+    winrt::hstring const& scale) const
+{
+    winrt::SnapPointBase winrtSnapPoint = safe_cast<winrt::SnapPointBase>(m_snapPoint);
+    SnapPointBase* snapPoint = winrt::get_self<SnapPointBase>(winrtSnapPoint);
+
+    return snapPoint->CreateRestingPointExpression(
+        compositor,
+        target,
+        scale);
+}
+
+template<typename T>
+winrt::ExpressionAnimation SnapPointWrapper<T>::CreateConditionalExpression(
+    winrt::Compositor const& compositor,
+    winrt::hstring const& target,
+    winrt::hstring const& scale) const
+{
+    winrt::SnapPointBase winrtSnapPoint = safe_cast<winrt::SnapPointBase>(m_snapPoint);
+    SnapPointBase* snapPoint = winrt::get_self<SnapPointBase>(winrtSnapPoint);
+
+    return snapPoint->CreateConditionalExpression(
+        m_actualApplicableZone,
+        compositor,
+        target,
+        scale);
+}
+
+template<typename T>
+void SnapPointWrapper<T>::DetermineActualApplicableZone(
+    SnapPointWrapper<T>* previousSnapPointWrapper,
+    SnapPointWrapper<T>* nextSnapPointWrapper)
+{
+    winrt::SnapPointBase winrtSnapPoint = safe_cast<winrt::SnapPointBase>(m_snapPoint);
+    SnapPointBase* snapPoint = winrt::get_self<SnapPointBase>(winrtSnapPoint);
+    SnapPointBase* previousSnapPoint = nullptr;
+    SnapPointBase* nextSnapPoint = nullptr;
+
+    if (previousSnapPointWrapper)
+    {
+        winrt::SnapPointBase winrtPreviousSnapPoint = safe_cast<winrt::SnapPointBase>(previousSnapPointWrapper->SnapPoint());
+        previousSnapPoint = winrt::get_self<SnapPointBase>(winrtPreviousSnapPoint);
+    }
+
+    if (nextSnapPointWrapper)
+    {
+        winrt::SnapPointBase winrtNextSnapPoint = safe_cast<winrt::SnapPointBase>(nextSnapPointWrapper->SnapPoint());
+        nextSnapPoint = winrt::get_self<SnapPointBase>(winrtNextSnapPoint);
+    }
```

Comment:
I prefer make the null check and get the actual snappoint as a function. then you can use it like this
if (auto snappoint = GetInnerSnapPoint(nextSnapPointWrapper))
{
}

Reply:
I added a SnapPointWrapper::GetSnapPointFromWrapper static method.